### PR TITLE
Category filtering for NordVPN

### LINKED ISSUE #1643: Feature request: Filter NordVPN servers by group
### What's the feature 🧐

NordVPN servers have groups, it would be nice to be able to filter by group.

- [ ] Change `internal/providers/nordvpn/updater` code
- [ ] Update `internal/storage/servers.json`
- [ ] Update Github wiki NorVPN servers list

Blocked by #1380 

### Extra information and references

Requested in:

- https://github.com/qdm12/gluetun/issues/719#issuecomment-1258403893
- https://github.com/qdm12/gluetun/issues/1541

### PR DESCRIPTION
**Pull request by AdamHebby**
Aims to fix #1643 

I am _not_ a GO developer, please read this code carefully!

image

image

After talking to NordVPN support, as long as a server supports P2P, a P2P connection will be established once it's determined there is a need for it (IE seeing P2P traffic). There's no way of knowing if it has connected to the P2P server other than contacting support with the IP address or results from `curl https://ipleak.net/json/`. I did ask if there was a way of starting the connection and specifying P2P from the get-go to avoid any switching but they said there was no documentation on this.

I've also [seen online](https://www.reddit.com/r/nordvpn/comments/q6astb/comment/hgbawxq/) that this switch from standard to P2P can result in you being reconnected to a "better" server, but this could also introduce latency.

TL;DR Nord claims it "just works"

Also unsure on testing? I've added a single test case for the categories - please let me know if there's anything more I can do here.

### CODE REVIEW

**Comment by AdamHebby** [comment]
@qdm12 Can I get your review on this please? Also I saw you were concerned about the potential file size increase, it's rather minimal at .9MB increase.

File: internal/provider/nordvpn/updater/models.go [hunk 1]
```diff
@@ -83,6 +85,18 @@ func (s *serverData) region() (region string) {
 	return ""
 }
 
+// categories returns the list of categories for the server.
+func (s *serverData) categories() (categories []string) {
+	cats := make([]string, 0, len(s.Groups))
+
```
Comment (qdm12) [inline internal/provider/nordvpn/updater/models.go:91]: nit just for the sake of using the returned variable and not create another 😉 

```go
	categories = make([]string, 0, len(s.Groups))
```

You would have to also update `cats` -> `categories`

File: internal/provider/nordvpn/updater/models.go [hunk 2]
```diff
@@ -83,6 +85,18 @@ func (s *serverData) region() (region string) {
...
+	for _, group := range s.Groups {
+		if group.Type.Identifier == "legacy_group_category" {
+			cats = append(cats, group.Identifier)
+		}
```
Comment (qdm12) [inline internal/provider/nordvpn/updater/models.go:95]: Quick question: why do we only take legacy group categories? Is there a new group category we can use instead?

File: internal/provider/nordvpn/updater/models.go [hunk 3]
```diff
@@ -47,6 +47,8 @@ type serverData struct {
 		} `json:"metadata"`
 	} `json:"technologies"`
 	Groups []struct {
+		// Identifier can be 'legacy_standard', 'legacy_p2p'.
+		Identifier string `json:"identifier"`
```
Comment (qdm12) [inline internal/provider/nordvpn/updater/models.go:51]: If the group can only be `legacy_standard` or `legacy_p2p`, would it be reasonable instead to just set the `PortForward` server boolean when encountering `legacy_p2p` and not have categories at all?

File: internal/provider/utils/filtering.go [hunk 4]
```diff
@@ -97,3 +101,19 @@ func filterByPossibilities[T string | uint16](value T, possibilities []T) (filte
 	}
 	return true
 }
+
+func filterMultipleByPossibilities[T string | uint16](values []T, possibilities []T) (filtered bool) {
```
Comment (qdm12) [inline internal/provider/utils/filtering.go:105]: nit since this is only used for []string (categories), no need to have generic for now:

```suggestion
func filterMultipleByPossibilities(values, possibilities []string) (filtered bool) {
```

File: internal/provider/utils/filtering.go [hunk 5]
```diff
@@ -97,3 +101,19 @@ func filterByPossibilities[T string | uint16](value T, possibilities []T) (filte
...
+
+	pass := true
+	for _, value := range values {
+		if !filterByPossibilities(value, possibilities) {
+			pass = false
+			break
+		}
+	}
+
+	return pass
```
Comment (qdm12) [inline internal/provider/utils/filtering.go:118]: Correct me if I'm wrong, but I think this is meant to be:

```suggestion
	for _, value := range values {
		filtered = filterByPossibilities(value, possibilities)
		if filtered {
			return true
		}
	}

	return false
```

File: internal/storage/filter.go [hunk 6]
```diff
@@ -119,6 +123,22 @@ func filterByPossibilities[T string | uint16](value T, possibilities []T) (filte
...
+	for _, value := range values {
+		if !filterByPossibilities(value, possibilities) {
+			pass = false
+			break
+		}
+	}
+
+	return pass
+}
+
```
Comment (qdm12) [inline internal/storage/filter.go:141]: same comment as for the other filterMultipleByPossibilities definition 😉

**Review by qdm12 — COMMENTED** [review]
Nice work 👍 💯 

I left a few comments, the most important one being:

> If the group can only be `legacy_standard` or `legacy_p2p`, would it be reasonable instead to just set the `PortForward` server model boolean to true when encountering `legacy_p2p` and not have categories at all?

That of course removes a lot of the fun in this PR, but it does make sense so far to me; even if there are other categories, I'm not certain there would be a usage to filter by category except for p2p? 🤔

File: internal/provider/nordvpn/updater/models.go (hunk 1 above)
Comment (AdamHebby) [reply internal/provider/nordvpn/updater/models.go:91]: I'm not sure what you mean here? Are you saying to rename cats to categories?
image

If I were to then remove the `make` it would add dupes to categories. Or are you saying to instead remove from categories and invert the IF logic?

File: internal/provider/nordvpn/updater/models.go (hunk 3 above)
Comment (AdamHebby) [reply internal/provider/nordvpn/updater/models.go:51]: Just double checked and turns out I missed `legacy_dedicated_ip`
```
"legacy_dedicated_ip"
"legacy_p2p"
"legacy_standard"
```

NGL I'm not sure what you mean about the PortForward bit, what relation to this does it have?

Side note: NordVPN does not support port forwarding

File: internal/provider/utils/filtering.go (hunk 4 above)
Comment (AdamHebby) [reply internal/provider/utils/filtering.go:105]: Should I also update `filterByPossibilities` since that's what I copied?

File: internal/provider/utils/filtering.go (hunk 5 above)
Comment (AdamHebby) [reply internal/provider/utils/filtering.go:118]: image

It's kinda based on this, but from my understanding, a TRUE evaluation means to filter out the value? Pls read filtering.go since you'll understand it more than me and advise pls

File: internal/provider/nordvpn/updater/models.go (hunk 3 above)
Comment (qdm12) [reply internal/provider/nordvpn/updater/models.go:51]: Yes my bad I may had mixed with another provider.
Another question: wasn't the server groups meant to be for example `Double_VPN`, `Africa_The_Middle_East_And_India`, etc. as mentioned in https://github.com/qdm12/gluetun/issues/1541#issue-1680084182 ?

File: internal/provider/nordvpn/updater/models.go (hunk 1 above)
Comment (qdm12) [reply internal/provider/nordvpn/updater/models.go:91]: No just have it as:

```go
func (s *serverData) categories() (categories []string) {
	categories = make([]string, 0, len(s.Groups))
```

and that will work 😉 `:=` is used to declare new variables, but `categories` is already there as the return variable.
If you use `make([]string, 0, len(s.Groups))`, the capacity of the slice is len(s.Groups), but the length is 0, so appending will start at index 0, then 1 etc.

File: internal/provider/nordvpn/updater/models.go (hunk 2 above)
Comment (qdm12) [reply internal/provider/nordvpn/updater/models.go:95]: Ok! Should we use the `title` instead since it's a bit more user friendly?

**Review by qdm12 — COMMENTED** [review]
Sorry for the delay re-reviewing, here are some answers on previous conversations! 👍

File: internal/provider/utils/filtering.go (hunk 5 above)
Comment (qdm12) [reply internal/provider/utils/filtering.go:118]: You're correct in your understanding of `filterByPossibilities`, but your function returns "do not filter" as soon as any value from `values` is correct; we want to check all values are correct. I'm fixing this 😉

File: internal/provider/nordvpn/updater/models.go (hunk 2 above)
Comment (qdm12) [reply internal/provider/nordvpn/updater/models.go]: Using the title, so now it's one of: Dedicated IP, Double VPN, Onion Over VPN, P2P, Standard VPN servers

**Comment by qdm12** [comment]
Actually never mind, checking more the response from https://api.nordvpn.com/v2/servers?limit=10

it shows `Legacy category` for anything not `regions`, so let's keep it with categories.
